### PR TITLE
transport: deny incoming peer certs with wrong IP SAN

### PR DESCRIPTION
**Pull request by heyitsanthony**
/cc @hongchaodeng

### CODE REVIEW

File: pkg/transport/listener_tls.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+			}
+
+			st := tlsConn.ConnectionState()
+			if len(st.PeerCertificates) > 0 {
+				cert := st.PeerCertificates[0]
+				if len(cert.IPAddresses) > 0 || len(cert.DNSNames) > 0 {
+					addr := tlsConn.RemoteAddr().String()
+					h, _, herr := net.SplitHostPort(addr)
+					if herr != nil || cert.VerifyHostname(h) != nil {
+						return
```
Comment (xiang90) [inline pkg/transport/listener_tls.go]: what error the other side receives if we close the conn here?

Comment (heyitsanthony) [inline pkg/transport/listener_tls.go]: EOF

Comment (xiang90) [inline pkg/transport/listener_tls.go]: this is no ideal. we probably need to log explicitly at server side.

there is no easy way to terminate it inside the TLS handshake path?

Comment (heyitsanthony) [inline pkg/transport/listener_tls.go]: If it did the verification, it'd be at https://golang.org/src/crypto/tls/handshake_server.go#L735 but there is no way to set `opts.DNSNames` which is needed to do the check at https://golang.org/src/crypto/x509/verify.go#L298. The go standard library even discourages doing this kind of check on the server side: https://golang.org/src/crypto/tls/conn.go#L1377

Comment (xiang90) [inline pkg/transport/listener_tls.go]: ok. let's log the error at server side. so that at least, human operator can know what is going on.

Comment (heyitsanthony) [inline pkg/transport/listener_tls.go]: Can't log with plog or it'll cause issues with clientv3 imports (9b8412773959d9e940c4319e88af23163e5c2fcf). Should there be an optional logging function in tls.Config then?

Comment (heyitsanthony) [inline pkg/transport/listener_tls.go]: s/tls.Config/transport.TLSInfo

Comment (xiang90) [inline pkg/transport/listener_tls.go]: add an interface or func seems ok.

**Comment by xiang90** [comment]
lgtm

**Comment by cxwen** [comment]
If the etcd nodes interact with each other through the domain name, only the domain name in the certificate does not have an ip. For example, deploy the etcd cluster in a k8s cluster, the following error will occur:

```
2019-04-04 16:48:02.534833 W | etcdserver: failed to reach the peerURL(https://xwcheng-ec-1.xwcheng-ec.xwcheng:2380) of member 2da6ba0f9279118b (Get https://xwcheng-ec-1.xwcheng-ec.xwcheng:2380/version: dial tcp 100.64.180.218:2380: getsockopt: connection refused)
2019-04-04 16:48:02.537128 W | etcdserver: cannot get the version of member 2da6ba0f9279118b (Get https://xwcheng-ec-1.xwcheng-ec.xwcheng:2380/version: dial tcp 100.64.180.218:2380: getsockopt: connection refused)
2019-04-04 16:48:02.616903 I | etcdmain: rejected connection from "100.64.180.218:44116" (error "tls: \"100.64.180.218\" does not match any of DNSNames [\"*.xwcheng-ec.xwcheng\" \"*.xwcheng-ec.xwcheng.cluster.local\"]", ServerName "xwcheng-ec-0.xwcheng-ec.xwcheng", IPAddresses [], DNSNames ["*.xwcheng-ec.xwcheng" "*.xwcheng-ec.xwcheng.cluster.local"])
```

How to solve this problem?

**Pull request merged**